Add OpenRouter & Gemini providers (streaming + list_models) + examples

srimon12 opened the pull request:
### Summary

Adds **OpenRouter** and **Gemini** providers with streaming and model-listing support.

### Changes

* **`openrouter_provider.py`** — Uses OpenAI SDK (custom base URL), supports streaming and `/api/v1/models`.
* **`gemini_provider.py`** — Uses `google-genai` SDK, supports non-streaming, streaming, and `models.list()`.
* **`core.py`** — Registers both providers.
* **Examples** — Added simple usage and streaming examples for each.

### Notes

Follows existing wrapper patterns; no core API changes.
Streaming prints live tokens and returns the final text, consistent with other providers.

anubhavgirdhar1 committed:
Merge pull request #32 from anubhavgirdhar1/dev

sync: dev to main

anubhavgirdhar1 committed:
chore: fixed versioning and deps

anubhavgirdhar1 committed:
fix: fixed sum stupid fkall bugs, ollama was trppin fr, fixed menu by adding a static path thatll remain constant for the user, needs more testing though, added ascii art to disable logs, removed fkin verisoning that was disaster, need logic here, my dih wrote this code

anubhavgirdhar1 committed:
fix: changed model name in example files

Commit:
fix: minor fixes in bedrock provider, need to think of a new logic to put in bedrock creds

poojashree443 committed:
feat(providers): add OpenRouter & Gemini (streaming + list_models) and examples

poojashree443 committed:
feat(providers): add OpenRouter & Gemini (streaming + list_models) and examples

srimon12 committed:
Merge branch 'dev' into add-openrouter-gemini

anubhavgirdhar1 commented on wrapper/config.py:
```diff
@@ -1 +1 @@
-SHOW_LOGS = False
```
@srimon12 lets keep it false by default

anubhavgirdhar1 commented on wrapper/providers/gemini_provider.py:
```diff
@@ -0,0 +1,156 @@
+log = ColorLogger(enable_debug=SHOW_LOGS)
+
+
+def _messages_to_contents(messages: List[Dict[str, Any]]) -> List[types.Content]:
```
@srimon12 I did not fully understand why do we need this function. Do we really need this?

srimon12 replied:
Yes, this helper is needed. it converts our OpenAI-style messages format into the Gemini SDK’s contents format. It handles small differences like role mapping (assistant > model) and merges any system messages into the first user message. This keeps the generate() and generate_stream() methods short and avoids repeating that logic in both places.

anubhavgirdhar1 commented on wrapper/providers/gemini_provider.py:
```diff
@@ -0,0 +1,156 @@
+# wrapper/providers/gemini_provider.py
+
+import sys
```
@srimon12 I was thinking to remove sys usage all together, since this just kills the stream as soon as it is either finished or in error state. let me know if you think other wise.

srimon12 replied:
removed all sys usage.

anubhavgirdhar1 commented on wrapper/providers/gemini_provider.py:
```diff
@@ -0,0 +1,156 @@
+                )
+                return (resp.text or "").strip()
+
+            output_chunks: List[str] = []
```
@srimon12  - Yielding would be the standard way. Lets change that as well. Let me know if you think other wise.

srimon12 replied:
generate_stream() now yields text chunks directly.

anubhavgirdhar1 commented on wrapper/providers/openrouter_provider.py:
```diff
@@ -0,0 +1,142 @@
+        self.default_system_prompt = "You are a helpful AI assistant."
+        self.base_api_url = "https://openrouter.ai/api/v1"
+
+        # Precompute optional extra headers for leaderboard attribution
```
@srimon12 - Just curious, what would be use for this?

srimon12 replied:
Removed

anubhavgirdhar1 commented on wrapper/providers/openrouter_provider.py:
```diff
@@ -0,0 +1,142 @@
+                    extra_headers=self.extra_headers if self.extra_headers else None,
+                    **kwargs
+                )
+                for chunk in stream_resp:
```
@srimon12 - Yeild here too

srimon12 replied:
Done

anubhavgirdhar1 commented on examples/gemini_stream_ex.py:
@srimon12 - lets have both stream and non stream example in the same file, what do you think?

srimon12 replied:
merged both stream and non-stream examples into a single file for each provider (OpenRouter + Gemini).

anubhavgirdhar1 reviewed (commented):
@srimon12 @poojashree443 Left some comments, let me know if you feel otherwise.

srimon12 committed:
chore: address review feedback (yield streaming, no sys, minimal helper, merged examples)

srimon12 commented:
Addressed all review points:
1. Removed sys usage.
2. generate_stream() yields chunks; generate(stream=True) returns assembled text.
3. Dropped extra OpenRouter headers.
4. Kept minimal _messages_to_contents() helper with clear inline comment.
5. Combined stream + non-stream examples.

anubhavgirdhar1 commented on wrapper/providers/gemini_provider.py:
@srimon12 - You have defined two methods, for stream and non stream. Instead you can define one single method and can have a check for the 'stream' arg. This would make it consistent and compatible to existing core class.

srimon12 replied:
done streamline both method to single one..

anubhavgirdhar1 commented on wrapper/providers/openrouter_provider.py:
@srimon12 - Same goes for this, we would need single method, with a check of stream arg.

anubhavgirdhar1 reviewed (commented):
@srimon12 - Left comments.

anubhavgirdhar1 commented:
@anshuman-parida  - Can you test both OpenRouter and Gemini please.

Commit:
chore: Standardize and harden Gemini, OpenRouter, and Groq

This commit refactors the Gemini, OpenRouter, and Groq providers to follow a single, robust, and maintainable pattern.

Key changes:
- Consolidates all logic within the `generate` method by inlining previous helper functions, which simplifies the code structure.
- Implements consistent, generator-based streaming across all three providers and removes side effects like printing to stdout.
- Replaces generic exception handling with specific, documented SDK errors for far more informative logging and resilience.
- Restores all original `generate` parameters to the Groq provider to ensure a consistent API.

srimon12 commented:
Pushed updates to standardize the Gemini, OpenRouter, and Groq providers.

They now all share the same clean structure:

1. Inlined generate() logic, removing helper methods.

2. Consistent, generator-based streaming.

3. Robust and specific SDK error handling for better logs.

This makes all three providers much more consistent and maintainable.

anubhavgirdhar1 commented on examples/list_models.py:
@srimon12 - love this, can you move it in the examples folder.

anubhavgirdhar1 commented on test.py:
@srimon12 - add this in gitignore as well, since people mostly do this.

anubhavgirdhar1 reviewed (commented):
Looks good. @anshuman-parida - need to you test this.

srimon12 committed:
chore: moved list_models.py and update gitignore

srimon12 commented:
Done. Pls check and test .

anubhavgirdhar1 commented:
This looks good. Tested and Merging the PR.
